scipy/stats/qmc.py

```diff
@@ -0,0 +1,648 @@
+"""Quasi-Monte Carlo methods.
+
+Define function to generate sample of points in the unit hypercube.
+
+"""
+
+from __future__ import division
+
+import copy
+import numpy as np
+from scipy.optimize import brute
+from scipy._lib._util import check_random_state
+
+try:
+    from scipy.optimize import basinhopping
+
+    have_basinhopping = True
+except ImportError:
+    have_basinhopping = False
+
+
+def discrepancy(sample, bounds=None, iterative=False):
+    """Discrepancy.
+
+    Compute the centered discrepancy on a given sample.
+    It is a measure of the uniformity of the points in the parameter space.
+    The lower the value is, the better the coverage of the parameter space is.
+
+    Parameters
+    ----------
+    sample : array_like (n_samples, k_vars)
+        The sample to compute the discrepancy from.
+    bounds : tuple or array_like ([min, k_vars], [max, k_vars])
+        Desired range of transformed data. The transformation apply the bounds
+        on the sample and not the theoretical space, unit cube. Thus min and
+        max values of the sample will coincide with the bounds.
+    iterative : bool
+        Must be False if not using it for updating the discrepancy.
+
+    Returns
+    -------
+    discrepancy : float
+        Centered discrepancy.
+
+    References
+    ----------
+    [1] Fang et al. "Design and modeling for computer experiments",
+      Computer Science and Data Analysis Series Science and Data Analysis
+      Series, 2006.
+
+    """
+    sample = np.asarray(sample)
+
+    n_samples, dim = sample.shape
+
+    if iterative:
+        n_samples += 1
+
+    # Sample scaling from bounds to unit hypercube
+    if bounds is not None:
+        min_ = bounds.min(axis=0)
+        max_ = bounds.max(axis=0)
+        sample = (sample - min_) / (max_ - min_)
+
+    abs_ = abs(sample - 0.5)
+    disc1 = np.sum(np.prod(1 + 0.5 * abs_ - 0.5 * abs_ ** 2, axis=1))
+
+    prod_arr = 1
+    for i in range(dim):
+        s0 = sample[:, i]
+        prod_arr *= (1 +
+                     0.5 * abs(s0[:, None] - 0.5) + 0.5 * abs(s0 - 0.5) -
+                     0.5 * abs(s0[:, None] - s0))
+    disc2 = prod_arr.sum()
+
+    c2 = ((13.0 / 12.0) ** dim - 2.0 / n_samples * disc1 +
+          1.0 / (n_samples ** 2) * disc2)
+
+    return c2
+
+
+def update_discrepancy(x_new, sample, initial_disc, bounds=None):
```

Comment:
this is not public but also not used from another function. can it be removed?

Reply:
This could be use to increase the size of an existing design. If we want to optimize the discrepancy then using this update function is faster.